enrollment-server-onboarding-provider-innovatrics/src/main/java/com/wultra/app/onboardingserver/provider/innovatrics/InnovatricsDocumentVerificationProvider.java

```diff
@@ -55,31 +120,271 @@ public boolean shouldStoreSelfie() {
+            results.setStatus(DocumentVerificationStatus.ACCEPTED);
+        }
+        results.setVerificationId(UUID.randomUUID().toString());
+        return results;
     }
 
     @Override
     public DocumentsVerificationResult getVerificationResult(OwnerId id, String verificationId) throws RemoteCommunicationException, DocumentVerificationException {
-        return null;
+        logger.warn("Unexpected state of documents with verificationId={}, {}", verificationId, id);
+        throw new UnsupportedOperationException("Method getVerificationResult is not supported by Innovatrics provider.");
     }
 
     @Override
     public Image getPhoto(String photoId) throws RemoteCommunicationException, DocumentVerificationException {
-        return null;
+        logger.warn("Unexpected document portrait query for customerId={}", photoId);
+        throw new UnsupportedOperationException("Method getPhoto is not implemented by Innovatrics provider.");
     }
 
     @Override
     public void cleanupDocuments(OwnerId id, List<String> uploadIds) throws RemoteCommunicationException, DocumentVerificationException {
-
+        // Pages of the same document have same uploadId (= customerId), no reason to call delete for each one.
+        final List<String> distinctUploadIds = uploadIds.stream().distinct().toList();
+        logger.info("Invoked cleanupDocuments, {}", id);
+        for (String customerId : distinctUploadIds) {
+            innovatricsApiService.deleteCustomer(customerId, id);
+        }
     }
 
     @Override
     public List<String> parseRejectionReasons(DocumentResultEntity docResult) throws DocumentVerificationException {
-        return null;
+        logger.debug("Parsing rejection reasons of {}", docResult);
+        final String rejectionReasons = docResult.getRejectReason();
+        if (!StringUtils.hasText(rejectionReasons)) {
+            return Collections.emptyList();
+        }
+
+        return deserializeFromString(rejectionReasons);
     }
 
     @Override
     public VerificationSdkInfo initVerificationSdk(OwnerId id, Map<String, String> initAttributes) throws RemoteCommunicationException, DocumentVerificationException {
-        return null;
+        logger.debug("#initVerificationSdk does nothing for Innovatrics, {}", id);
+        return new VerificationSdkInfo();
+    }
+
+    /**
+     * Create a new customer resource.
+     * @param ownerId owner identification.
+     * @return ID of the new customer.
+     * @throws RemoteCommunicationException if the resource was not created properly.
+     */
+    private String createCustomer(final OwnerId ownerId) throws RemoteCommunicationException {
+        return innovatricsApiService.createCustomer(ownerId)
+                .map(CreateCustomerResponse::getId)
+                .filter(StringUtils::hasText)
+                .orElseThrow(() -> new RemoteCommunicationException("A customer resource could not be created, %s".formatted(ownerId)));
```

Comment:
🤔 The remote server has responded, so I am hesitating to use `DocumentVerificationException` instead of `RemoteCommunicationException`. I believe that `Provider` is a business logic layer. Communication exception should start in `InnovatricsApiService`, shouldn`t it?